examples/extra_element_integer/clustering_algorithm.h

```diff
@@ -50,54 +50,90 @@ void AddVariablesToSystem(libMesh::EquationSystems &equation_systems,
-  dof_map.dof_indices(&elem, dof_indices, variable_num);
+  dof_map.dof_indices(elem, dof_indices, variable_num);
 
   std::vector<double> solution_value(1);
   system.solution->get(dof_indices, solution_value);
 
   return static_cast<double>(solution_value[0]);
 }
 
-const unsigned int FindCluster(libMesh::Mesh &mesh,
-                               LinearImplicitSystem &system,
-                               const std::string &variable_name) {
+void SetElementIDasClusterID(libMesh::Mesh &mesh, unsigned int index) {
+  for (const auto &elem : mesh.element_ptr_range()) {
+    elem->set_extra_integer(index, elem->id());
+  }
+}
 
-  const unsigned int variable_num = system.variable_number(variable_name);
+bool BelongToCluster(libMesh::Elem *elem, libMesh::Elem *neighbor_elem,
+                     int variable_num, libMesh::LinearImplicitSystem &system) {
 
-  const unsigned int index = mesh.add_elem_integer(variable_name);
+  double element_solution = GetElementDataFromMesh(system, elem, variable_num);
+  double neighbor_solution =
+      GetElementDataFromMesh(system, neighbor_elem, variable_num);
+  if (static_cast<int>(element_solution) ==
+      static_cast<int>(neighbor_solution)) {
 
-  for (const auto &elem : mesh.element_ptr_range()) {
-
-    int element_solution =
-        static_cast<int>(GetElementDataFromMesh(system, *elem, variable_num));
+    return true;
+  } else {
+    return false;
+  }
+}
 
-    bool belong_to_a_cluster = false;
-    for (unsigned int side = 0; side < elem->n_sides(); ++side) {
-      const Elem *neighbor = elem->neighbor_ptr(side);
-      if (neighbor) {
+const unsigned int FindCluster(libMesh::Mesh &mesh,
+                               libMesh::LinearImplicitSystem &system,
+                               const std::string &variable_name) {
+  const unsigned int variable_num = system.variable_number(variable_name);
+  const unsigned int index = mesh.add_elem_integer(variable_name);
 
-        int neighbor_element_solution = static_cast<int>(
-            GetElementDataFromMesh(system, *neighbor, variable_num));
+  SetElementIDasClusterID(mesh, index);
 
-        if (element_solution == neighbor_element_solution) {
-          belong_to_a_cluster = true;
-          // Danger Zone for future ref
-          // cluster_id should never be equal to element_solution
-          unsigned int cluster_id = element_solution;
-          elem->set_extra_integer(index, cluster_id);
-          break;
+  for (const auto &elem : mesh.element_ptr_range()) {
+    if (elem->get_extra_integer(index) == elem->id()) {
+      for (unsigned int side = 0; side < elem->n_sides(); side++) {
+        libMesh::Elem *neighbor_elem = elem->neighbor_ptr(side);
+        if (neighbor_elem) {
+          if (BelongToCluster(elem, neighbor_elem, variable_num, system)) {
+
+            std::vector<libMesh::Elem *> neighbor_list;
+            neighbor_elem->set_extra_integer(index, elem->id());
+            neighbor_list.push_back(neighbor_elem);
```

Comment:
If you are careful, this can all be done by the similar code below.  You don't need a special case for this element that is different from its neighbors.